### PR TITLE
Enable cors for for proxy routes

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes https://github.com/microsoft/reverse-proxy/issues/285

### CODE REVIEW

**Comment by Tratcher** [comment]
That was fast 😉

File: src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs [hunk 1]
```diff
@@ -177,5 +184,40 @@ private async Task<bool> ValidateAuthorizationPolicyAsync(string authorizationPo
...
+            }
+
+            if (string.Equals(CorsConstants.Disable, corsPolicyName, StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+
+            try
+            {
+                var policy = await _corsPolicyProvider.GetPolicyAsync(_dummyHttpContext, corsPolicyName);
```
Comment (Kahbazi) [inline src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs:207]: I'm not sure if I did the right thing here by passing a dummy `HttpContext` to `GetPolicyAsync`. The default implementation ([DefaultCorsPolicyProvider](https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/CORS/src/Infrastructure/DefaultCorsPolicyProvider.cs)) is not using HttpContext at all.

File: src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs [hunk 2]
```diff
@@ -40,6 +43,7 @@ internal class RouteValidator : IRouteValidator
             DnsLabelRegexPattern +
             @"$";
         private static readonly Regex _hostNameRegex = new Regex(HostNameRegexPattern);
+        private static readonly HttpContext _dummyHttpContext = new DefaultHttpContext();
```
Comment (Tratcher) [inline src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs:46]: Don't make a static context, even if it's a dummy. The type isn't thread safe. Create a local instance in ValidateCorsPolicyAsync.

File: src/ReverseProxy/Abstractions/RouteDiscovery/Contract/ProxyRoute.cs [hunk 3]
```diff
@@ -39,6 +39,14 @@ public sealed class ProxyRoute : IDeepCloneable<ProxyRoute>
         /// </summary>
         public string AuthorizationPolicy { get; set; }
 
+        /// <summary>
+        /// The name of the CorsPolicy to apply to this route.
+        /// If not set then the route won't be matched for cors preflight request.
+        /// Set to "Default" to enable cors with the default policy.
+        /// Set to "Disable" to disable cors checks for this route.
+        /// </summary>
+        public string CorsPolicy { get; set; }
```
Comment (Kahbazi) [inline src/ReverseProxy/Abstractions/RouteDiscovery/Contract/ProxyRoute.cs:48]: If I understand the routing correctly, with `Disable`, the proxy endpoint would be selected and the `CorsMiddleware` would set `Status204NoContent`. But when `CorsPolicy` is empty, no endpoint would be selected and response would probably end up being a `Status404NotFound`.

Comment (Tratcher) [reply src/ReverseProxy/Abstractions/RouteDiscovery/Contract/ProxyRoute.cs:48]: Yeah, I was looking at that. `Disable` doesn't turn off Cors processing, it refuses Cors requests.
https://github.com/dotnet/aspnetcore/blob/a729c4230eefd4b5c3e21f9496de608cd25a72b8/src/Middleware/CORS/src/Infrastructure/CorsMiddleware.cs#L144-L153
Please clarify the comment.

File: src/ReverseProxy/Abstractions/RouteDiscovery/Contract/ProxyRoute.cs [hunk 4]
```diff
@@ -39,6 +39,14 @@ public sealed class ProxyRoute : IDeepCloneable<ProxyRoute>
         /// </summary>
         public string AuthorizationPolicy { get; set; }
 
+        /// <summary>
+        /// The name of the CorsPolicy to apply to this route.
+        /// If not set then the route won't be matched for cors preflight request.
```
Comment (Tratcher) [inline src/ReverseProxy/Abstractions/RouteDiscovery/Contract/ProxyRoute.cs:44]:
```suggestion
        /// If not set then the route won't be automatically matched for cors preflight requests.
        /// Routes that match the Options method, or all methods, will still be able to process cors 
        /// preflight requests at the destination server.
```

File: src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs (hunk 1 above)
Comment (Tratcher) [reply src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs:207]: Providing a dummy makes sense, it doesn't allow nulls.

Comment (Kahbazi) [reply src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs:207]: Why `CorsPolicyProvider.GetPolicyAsync` needs a `HttpContext` in the first place, but `authorizationPolicyProvider.GetPolicyAsync` doesn't? 
Is there any chance that `CorsPolicyProvider.GetPolicyAsync(HttpContext,string)` could be obsoleted and  `CorsPolicyProvider.GetPolicyAsync(string)` added instead?

Comment (Tratcher) [reply src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs:207]: > Why `CorsPolicyProvider.GetPolicyAsync` needs a `HttpContext` in the first place

I don't know.

> Is there any chance that `CorsPolicyProvider.GetPolicyAsync(HttpContext,string)` could be obsoleted and `CorsPolicyProvider.GetPolicyAsync(string)` added instead?

Updating interfaces is one of the harder kinds of breaking changes to make. We'll still need to support 3.1 regardless.

**Comment by davidfowl** [comment]
This still requires the user to explicitly call UseCors right?

**Comment by Kahbazi** [comment]
> This still requires the user to explicitly call UseCors right?

Yes.

**Comment by Kahbazi** [comment]
When a route with cors is selected, is there any harm to send the cors headers to the cluster?

**Comment by Tratcher** [comment]
> When a route with cors is selected, is there any harm to send the cors headers to the cluster?

You mean should they be stripped off before forwarding? No. If someone wants to do that they can handle it in the transforms.

File: docs/docfx/articles/cors.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+```
+
+
+### DefaultPolicy
+
+Specifying the value `default` in a route's CORS parameter means that route will use the policy defined in [CorsOptions.DefaultPolicy](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.cors.infrastructure.corsoptions.defaultpolicyname).
+
+### Disable CORS
+
+Specifying the value `disable` in a route's CORS parameter means the CORS middleware will refuse the CORS requests.
```
Comment (halter73) [inline docs/docfx/articles/cors.md:81]: Nit:

```suggestion
Specifying the value `disable` in a route's `CorsPolicy` parameter means the CORS middleware will refuse the CORS requests.
```

Unless I'm misunderstanding and this is meant to say something else.

File: src/ReverseProxy/Service/ConfigModel/ParsedRoute.cs [hunk 6]
```diff
@@ -114,6 +122,11 @@ internal int GetConfigHash()
                 hash ^= AuthorizationPolicy.GetHashCode();
             }
 
+            if (!string.IsNullOrEmpty(CorsPolicy))
+            {
+                hash ^= CorsPolicy.GetHashCode();
```
Comment (pranavkm) [inline src/ReverseProxy/Service/ConfigModel/ParsedRoute.cs:127]: Why not use the `HashCode` type?

Comment (Tratcher) [reply src/ReverseProxy/Service/ConfigModel/ParsedRoute.cs:127]: That's new in 5.0, right? We still support 3.1.

Comment (pranavkm) [reply src/ReverseProxy/Service/ConfigModel/ParsedRoute.cs:127]: It's in 3.1: https://docs.microsoft.com/en-us/dotnet/api/system.hashcode?view=netcore-3.1

Comment (davidfowl) [reply src/ReverseProxy/Service/ConfigModel/ParsedRoute.cs:127]: It's not new in 5.0 it's in 3.1 but I was reserving that for when we rewrite all of this code.

Comment (Tratcher) [reply src/ReverseProxy/Service/ConfigModel/ParsedRoute.cs:127]: Yeah, don't worry about it in this PR.

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**